src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -225,9 +218,110 @@ private void pre20Upgrade() throws Exception {
                 }
             }
         }
-        if (hasCustomPre20HashFunction|| pre20UseType != null) {
+        if (hasCustomPre20HashFunction || pre20UseType != null) {
             logger.warn("Settings [{}] and [{}] are deprecated. Index settings from your old indices have been updated to record the fact that they "
                     + "used some custom routing logic, you can now remove these settings from your `elasticsearch.yml` file", DEPRECATED_SETTING_ROUTING_HASH_FUNCTION, DEPRECATED_SETTING_ROUTING_USE_TYPE);
         }
     }
+
+    /**
+     * Loads the current meta state for each index in the new cluster state and checks if it has to be persisted.
+     * Each index state that should be written to disk will be returned. This is only run for data only nodes.
+     * It will return only the states for indices that actually have a shard allocated on the current node.
+     *
+     * @param event           the cluster state event from which we figure out what is new in each index and should potentially be written
+     * @param currentMetaData the current index state in memory.
+     * @return iterable over all indices states that should be written to disk
+     */
+    public static Iterable<GatewayMetaState.IndexMetaWriteInfo> filterStateOnDataNode(ClusterChangedEvent event, MetaData currentMetaData) {
+        ClusterState newState = event.state();
+        ClusterState previousState = event.previousState();
+
+        Map<String, GatewayMetaState.IndexMetaWriteInfo> indicesToWrite = new HashMap<>();
+        RoutingNode newRoutingNode = newState.getRoutingNodes().node(newState.nodes().localNodeId());
+        RoutingNode previousRoutingNode = previousState.getRoutingNodes().node(newState.nodes().localNodeId());
```

Comment:
this one should go away. I see why it's here (because we only cache the currentMetaData which doesn't tell us anything about allocation) but now we have too ways of deciding what to do . One based on last written state and the other based on the last CS in the event. This is an issue. Can we cache a set of last written indices next to `currentMetaData` and use that?